### ISSUE TITLE
Remix Vite Plugin triggers watchers during Build

### ISSUE THREAD

**Issue by kube** [remix-run/remix#8313]
### Reproduction

I'm using the Vite Plugin for building an app (and deploy it to Production), and I have two issues:

1. `vite build` seems to trigger watchers on all files in my workspace.
   Using pnpm on Gitlab, I configured `viteConfig.server.watch.ignored` to include `**/.pnpm-store/**`.
   This prevents the CI to crash because of too many inotify watchers. 

   (See kinda related issue https://gitlab.com/gitlab-org/gitlab/-/issues/434329)

   > I'm not sure if the watchers are due to the Remix Plugin or to Vite itself, but I have to put context here.
   >
   > I don't understand why `vite build` would need to watch anything on the workspace?

2. Until version `2.4.0` it was possible to configure which files had to be ignored, but since `2.4.0` (in this [commit](https://github.com/remix-run/remix/commit/787bb83c0a38297644a93803bf17295f3574e180)), the configuration of the Server does not overwrite User config but does an entire replace instead.

#### Configuration

```ts
export default defineConfig({
  server: { port: 3000, {
    watch: {
     ignored: ["**/.pnpm-store/**"] // This is not read by the plugin anymore
    }
  },
  plugins: [
    tsConfigPaths(),
    remix({
      ignoredRouteFiles: ["**/.*", "**/*.css", "**/*.test.{js,jsx,ts,tsx}"],
    }),
  ],
});
```

### System Info

```shell
(System is not relevant here, the issue is on Gitlab CI.)

  System:
    OS: macOS 14.2
    CPU: (10) arm64 Apple M1 Max
    Memory: 16.54 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.3.0 - ~/.nvm/versions/node/v21.3.0/bin/node
    Yarn: 1.22.21 - ~/.nvm/versions/node/v21.3.0/bin/yarn
    npm: 10.2.4 - ~/.nvm/versions/node/v21.3.0/bin/npm
    pnpm: 8.7.6 - ~/Library/pnpm/pnpm
  Browsers:
    Chrome: 120.0.6099.109
    Safari: 17.2
  npmPackages:
    @remix-run/dev: 2.4.0 => 2.4.0 
    @remix-run/eslint-config: 2.4.0 => 2.4.0 
    @remix-run/express: 2.4.0 => 2.4.0 
    @remix-run/node: 2.4.0 => 2.4.0 
    @remix-run/react: 2.4.0 => 2.4.0 
    @remix-run/server-runtime: 2.4.0 => 2.4.0 
    vite: 5.0.10 => 5.0.10
```

### Used Package Manager

pnpm

### Expected Behavior

1. Remix Vite plugin should not trigger watchers during build.

2. It should be possible to configure which files should be ignored using `viteConfig.server.watch.ignored`.

### Actual Behavior

1. Remix Vite Plugin triggers watchers for all files in workspace (except `node_modules/` and `.git/` which are Vite defaults)

2. `viteConfig.server.watch.ignored` is entirely skipped by the Remix Vite Plugin.

**Comment by kube** [comment]
I just took a quick look at how SvelteKit managed the build, and it seems that they don't use `vite.createServer` but just `vite.build`:

https://github.com/sveltejs/kit/blob/c0f3c403b774df72b113646559e915e4f53a8566/packages/kit/src/exports/vite/index.js#L740

**Comment by kube** [comment]
Quick and dirty fix (until implementation using vite.build) could be to set `viteConfig.server.watch` to `null`.

**Comment by hi-ogawa** [comment]
Interesting, thanks for reporting the issue.
The code you're referencing (around `viteChildCompiler = await vite.createServer(...)`) is currently necessary for both `dev` and `build` and it's used for transforming code independently from Vite (but still benefiting from Vite-powered transform).

Note that Remix makes use of `vite.build` API for main build already, but `viteChildCompiler` is still necessary and I think this is somewhat internal detail (and thus not inheriting all user config via `{ server: { ... viteUserConfig.server } }` feels still right to me):
https://github.com/remix-run/remix/blob/9edcf88d3c25cdabbe628782927cae2c735242c6/packages/remix-dev/vite/build.ts#L71

To fix this, `viteConfig.server.watch: null` makes sense and I think that should be safe at least for build case. Thanks for the suggestion!

**Comment by markdalgleish** [comment]
Fixed by #8320.
